### PR TITLE
Implementing Linear Regression Using Gradient Descent in Python

### LINKED ISSUE #1165: Implementing Linear Regression using Gradient Descent
### Brief Summary: 
Describe the what, why, and how of your content idea in 2-5 sentences.
In the previous article, we derived the mathematical equations to find the line of best fit iteratively. In this article, I will code linear regression from scratch using numpy, and the equations derived earlier.
### Key Takeaways: 
What are the 3-5 most important things the reader should understand or be able to do after reading this article?
1. An overview of the mathematical equations.
2. Code the linear regression class
3. Define the various methods such as optimize, initialize_weights_and_bias, _computeError, optimize, etc., to perform computations as seen in the mathematical equations. 
4. Test the algorithm on the Boston housing dataset and compare the performance with the sklearn library.
5. Going through this article, one will understand how to convert mathematical equations to Python code using basic libraries such as NumPy. 
### References: 
Please list links to any published content/research that you intend to use to support/guide this article. (If none, please indicate 'N/A'.)
N/A

### PR DESCRIPTION
**Pull request by lalith1403**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.

**NOTE: (1st time contributors - submit ONLY one article to get familiar with the process and style-once published, you can then submit any number of articles.)**
      
#### Grammar Checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical Checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution Guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

this fixes #1165

### CODE REVIEW

**Comment by hectorkambow** [comment]
👋 Thank you for submitting this PR - we will review it ASAP - all PRs go into our queue will be looked over before being approved to be published.

File: articles/linear-regression-code/index.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
+In the previous article, we looked at the theory behind linear regression. In this article, we will implement the linear regression algorithm from scratch and understand the various steps involved. In my earlier article, I used the in-built function in scikit-learn to predict the houses' prices. Towards the end of the article, we will compare the two approaches and reason the outputs.
```
Comment (cyberShaw) [inline articles/linear-regression-code/index.md:1]: Could you add a link to the previous article?

File: articles/linear-regression-code/index.md [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+We get a training accuracy of about 71%, and test accuracy stands at 65%. A fun exercise would be to set `normalize` to `False` and try the same code. Check the train and test accuracies. You will understand the significance of normalization.
+
+```bash
+Train Score: 0.7142761537090165
+Test Score:  0.6504432415059116
+```
+
+### Comparison with scikit learn
+
+The training and test accuracy obtained using the [library](/engineering-education/house-price-prediction/) stands at 93% and 79.29%, respectively. We conclude that the data requires some non-linearity to be introduced, and polynomial regression would probably work much better than linear regression.
```
Comment (cyberShaw) [inline articles/linear-regression-code/index.md:131]:
```suggestion
The training and test accuracy obtained using the [library](/engineering-education/house-price-prediction/) stand at 93% and 79.29%, respectively. We conclude that the data requires some non-linearity to be introduced, and polynomial regression would probably work much better than linear regression. 
```

File: articles/linear-regression-code/index.md [hunk 3]
```diff
@@ -0,0 +1,135 @@
+In the previous article, we looked at the theory behind linear regression. In this article, we will implement the linear regression algorithm from scratch and understand the various steps involved. In my earlier article, I used the in-built function in scikit-learn to predict the houses' prices. Towards the end of the article, we will compare the two approaches and reason the outputs.
+
+### Introduction
+
+As discussed in the previous article, linear regression is a type of supervised learning algorithm. It is used in many applications in the financial industry. First, let us understand the various functions needed to implement a linear regression class, to begin with the coding aspect. 
+
```
Comment (cyberShaw) [inline articles/linear-regression-code/index.md:6]: Suggestion: You could have an additional sub-heading where you define terms such as Bias, Weights, Loss, Independent Variable, Dependent Variable, etc. (Just a single line definition)

I am aware that this article is supposed to be a continuation of another article, but if a beginner to ML is reading this, a glossary would be really helpful for reference.

File: articles/linear-regression-code/index.md [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+8. `plot`: In this method, we plot the loss function versus the number of iterations.
+9. `score`: This function calculates the accuracy. The accuracy is computed using the following formula:
+    $$ accuracy = \frac{(y - \hat{y})^2}{\sum{_i}^n(y-\bar{y})^2}$$
+
+### Code 
+
+Let us put together the information we collected above and create the `Regressor` class. 
+
+```py
+import numpy as np
```
Comment (cyberShaw) [inline articles/linear-regression-code/index.md:26]: It would be a good idea to add comments to your code, thereby making it easier for a beginner to follow.

File: articles/linear-regression-code/index.md [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+We use the Boston housing dataset to test the performance of the model built. The code to call the Boston housing dataset and to train the model is given below.
+
+```py
+from sklearn.datasets import load_boston
+from sklearn.model_selection import train_test_split
+
+data = load_boston()
+X_train, X_test, y_train,y_test = train_test_split(data.data, data.target,test_size=0.1)
+print(f"X_train:{X_train.shape}\ny_train:{y_train.shape}")
```
Comment (cyberShaw) [inline articles/linear-regression-code/index.md:104]: Please add code comments here too :)

**Review by cyberShaw — CHANGES_REQUESTED** [review]
Great and crisp article, @lalith1403!

I've suggested some minor improvements to the article. Please implement those and we can move on to final review.

**Comment by hectorkambow** [comment]
plagiarism check done

**Comment by cyberShaw** [comment]
@lalith1403 I just went through the changes and they look great! Could you add the glossary change that I suggested and we can mark this ready for final review.

**Comment by lalith1403** [comment]
Regarding the glossary, there is a detailed article on the same, which i did mention in the prerequisites section. I did that to avoid redundancy.
